recommendation/widedeep/widedeep.py

```diff
@@ -0,0 +1,119 @@
+
+import bz2
+import math
+import pickle
+import collections
+
+import numpy as np
+import tensorflow as tf
+
+import utils.misc as utils
+
+
+class WideDeep:
+    """
+    A class providing facilities for preprocessing and postprocessing of ImageNet validation dataset.
+    """
+
+    def __init__(self, batch_size: int, config=None, runner=None, dataset_path=None, tfrecords_path=None):
+
+        if tfrecords_path is None:
+            env_var = "WIDEDEEP_TFRECORDS_PATH"
+            tfrecords_path = utils.get_env_variable(
+                env_var, f"Path to tfrecords path has not been specified with {env_var} flag")
+            self.tfrecords_path = tfrecords_path
+
+        self.available_instances = sum(1 for _ in tf.compat.v1.python_io.tf_record_iterator(
+            self.tfrecords_path))
+
+        self.dataset_path = dataset_path
+        self.batch_size = batch_size
+        self.no_of_batches = math.ceil(float(self.available_instances // self.batch_size))
+        self.features_list = self.get_features_list(config, runner, self.no_of_batches)
+
+        self.current_feature = 0
+        self.correct = 0
+
+        super().__init__()
+
+    def input_fn(self, shuffle):
+        """Generate an input function for the Estimator."""
+
+        numeric_feature_names = ["numeric_1"]
+        string_feature_names = ["string_1"]
+
+        full_features_names = numeric_feature_names + string_feature_names + ["label"]
+        feature_datatypes = [tf.io.FixedLenSequenceFeature([], tf.float32, default_value=0.0, allow_missing=True)] + \
+                            [tf.io.FixedLenSequenceFeature([], tf.int64, default_value=0, allow_missing=True)] + \
+                            [tf.io.FixedLenSequenceFeature([], tf.int64, default_value=0, allow_missing=True)]
+
+        def _parse_function(proto):
+            f = collections.OrderedDict(
+                zip(full_features_names, feature_datatypes))
+            parsed_features = tf.io.parse_example(proto, f)
+            parsed_feature_vals_num = [tf.reshape(
+                parsed_features["numeric_1"], shape=[-1, 13])]
+            parsed_feature_vals_str = [tf.reshape(
+                parsed_features["string_1"], shape=[-1, 2]) for i in string_feature_names]
+            parsed_feature_vals = parsed_feature_vals_num + parsed_feature_vals_str
+
+            # labels
+            parsed_feature_vals_label = [tf.reshape(parsed_features[i], shape=[-1]) for i in ["label"]]
+            parsed_feature_vals = parsed_feature_vals + parsed_feature_vals_label
+
+            return parsed_feature_vals
+
+        # Extract lines from input files using the Dataset API.
+        dataset = tf.data.TFRecordDataset([self.tfrecords_path])
+        if shuffle:
+            dataset = dataset.shuffle(buffer_size=20000)
+        dataset = dataset.batch(self.batch_size)
+        dataset = dataset.map(_parse_function, num_parallel_calls=28)
+        dataset = dataset.prefetch(self.batch_size * 10)
```

Comment:
some explanation here would be welcomed - why 28 num parallel calls? if this is arbitrary it would be reasonable to introduce some variable to change that if needed be. Same for prefetch, buffer size